Kudryashov/epic cart iteration3

KudryashovAlexander opened the pull request:
Эпик: Корзина
Архитектура: MVVM
Верстка: кодом
Таск-трекер: [тут](https://github.com/orgs/PracticumTeam3/projects/1/views/1) 
Декомпозиция: [md](https://github.com/PracticumTeam3/iOS-FakeNFT-StarterProject-Public/blob/kudryashov/epic_cart_iteration3/decomposition/CartFlowDecomposition.md)

KudryashovAlexander committed:
Добавлена вью успешной оплаты

KudryashovAlexander committed:
Добавлен АлертКонтроллер при неудачной оплате

KudryashovAlexander committed:
Созданы файлы для WebView

KudryashovAlexander committed:
Merge branch 'kudryashov/epic_cart_iteration2' into kudryashov/epic_cart_iteration3

KudryashovAlexander committed:
Добавлена реализация WebView

KudryashovAlexander committed:
Добавлен AlertController на случай пропадания интернета

KudryashovAlexander committed:
Добавлено фактические время в декомпозицию

AndreyOvchinnikovDev reviewed (commented):
привет, посмотрел ПР все круто, добавить нечего)

FanEm commented on FakeNFT/Scenes /Cart/CartView/CartTableViewViewModel.swift:
```diff
@@ -87,8 +92,27 @@ final class CartTableViewViewModel {
+            return
+        }
+        switch netWorkAlert {
+        case .fetchOrder, .fetchNFT ,.changeOrder:
```
Лишний пробел перед запятой и отсутствие пробела после
`.fetchOrder, .fetchNFT, .changeOrder`

KudryashovAlexander replied:
Поправил

FanEm commented on FakeNFT/Scenes /Cart/CartView/CartTableViewViewModel.swift:
```diff
@@ -87,8 +92,27 @@ final class CartTableViewViewModel {
+        }
+    }
+    
+    private func checkAlert(_ netWorkAlert: NetWorkAlert?) {
```
net**w**orkAlert

KudryashovAlexander replied:
Поправил

FanEm commented on FakeNFT/Scenes /Cart/CartWebView/CartWebViewController.swift:
```diff
@@ -0,0 +1,141 @@
+    
+    init(viewModel: CartWebViewModel) {
+        self.viewModel = viewModel
+        super .init(nibName: nil, bundle: nil)
```
лишний пробел

KudryashovAlexander replied:
Поправил

FanEm commented on FakeNFT/Scenes /Cart/CartWebView/CartWebViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+        progress = newValue
+        if newValue == 1.0 {
+            progressHUDIsActive = false
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
```
А зачем задержка?

KudryashovAlexander replied:
Без задержки progressView скрывается до того как дойти до конца. То есть не успевает отрисоваться на экране.

FanEm commented on FakeNFT/Scenes /Cart/CartWebView/CartWebViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+    }
+    
+    func getRequest() -> URLRequest? {
+        guard let url = URL(string: "https://yandex.ru/legal/practicum_termsofuse/") else { return nil }
```
Возможно, нужно в строки положить. И можно для `en` сделать `yandex.com`, а для `ru` - `yandex.ru`

KudryashovAlexander replied:
Попробовал открыть yandex.com, идет перенаправление на другую [ссылку](https://docs.tripleten.com/legal/terms_of_use.html)

Оставлю как есть

FanEm commented on FakeNFT/Scenes /Cart/PaymentMethodView/PaymentViewController.swift:
```diff
@@ -157,12 +159,54 @@ final class PaymentViewController: UIViewController {
+        let alert = UIAlertController(title: L.Cart.failurePay,
+                                      message: nil,
+                                      preferredStyle: .alert)
+        let actionCansel = UIAlertAction(title: L.Cart.cansel,
```
can**c**el?

KudryashovAlexander replied:
поправил)

FanEm commented on FakeNFT/Scenes /Cart/PaymentMethodView/PaymentViewController.swift:
```diff
@@ -157,12 +159,54 @@ final class PaymentViewController: UIViewController {
+    }
+    
+    private func showAlert() {
+        let alert = UIAlertController(title: L.Cart.failurePay,
```
А у тебя же есть `CartAlertPresenter`, мб стоит сделать универсальное решение для алертов? Можно пока забить, а после мержа заиспользовать мое, если понравится (или написать общее удобное для всех)

KudryashovAlexander replied:
Посмотрим тогда после мерджа

FanEm commented:
В ландшафте не скроллится. Можно либо вложить в скроллвью, либо сделать чтобы конкретно этот экран отображался только в портрете.
<img width="867" alt="Screenshot 2023-11-01 at 21 52 18" src="https://github.com/PracticumTeam3/iOS-FakeNFT-StarterProject-Public/assets/2598509/89862b88-2852-4562-be8e-c92c3ff3c93b">

KudryashovAlexander committed:
Поправлено после кросс-ревью

KudryashovAlexander commented:
Всем привет!
Спасибо за ревью.

Скролинг на экран удачной оплаты добавил.

Helvetii951 commented on FakeNFT/Scenes /Cart/CartWebView/CartWebViewController.swift:
```diff
@@ -0,0 +1,141 @@
+import UIKit
+import WebKit
+
+final class CartWebViewController: UIViewController {
+    
+    private let viewModel: CartWebViewModel
+    
+    private let webView: WKWebView = {
+        let view = WKWebView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        return view
+    }()
```
Здесь можно использовать [sfsafariviewcontroller](https://developer.apple.com/documentation/safariservices/sfsafariviewcontroller). В несколько строчек можно создать экземпляр, передать урл и отобразить веб-страницу. Плюс мы получим плюшки - почти все, что есть в Safari.

[WKWebView or SFSafariViewController](https://hub.packtpub.com/should-i-use-wkwebview-or-sfsafariviewcontroller-for-web-views-in-my-app-from-news-apple-developer/)

KudryashovAlexander replied:
Спасибо, буду иметь в виду.

Helvetii951 commented on FakeNFT/Scenes /Cart/SucessPayView/SuccessPayViewController.swift:
```diff
@@ -0,0 +1,122 @@
+        return imageView
+    }()
+
+    private let successLabel: UILabel = {
+        let label = UILabel()
+        label.text = L.Cart.successPay
+        label.textColor = A.Colors.blackDynamic.color
+        label.font = .bold22
+        label.textAlignment = .center
+        label.numberOfLines = 2
+        label.translatesAutoresizingMaskIntoConstraints = false
+        return label
+    }()
+    
+    private lazy var backCatalogeButton: UIButton = {
+        let button = UIButton()
+        button.setTitle(L.Cart.backCataloge, for: .normal)
+        button.setTitleColor(A.Colors.whiteDynamic.color, for: .normal) 
+        button.titleLabel?.font = .bold17
+        button.backgroundColor = A.Colors.blackDynamic.color
+        button.addTarget(self, action: #selector(backCataloge), for: .touchUpInside)
+        button.layer.masksToBounds = true
+        button.layer.cornerRadius = 16
+        button.translatesAutoresizingMaskIntoConstraints = false
+        return button
+    }()
```
Рекомендую чаще использовать [ленивые свойства ](https://abhimuralidharan.medium.com/lazy-var-in-ios-swift-96c75cb8a13a). В нашем случае это какого-то буста не даст. Но, если какая-то вьюха не будет использоваться, можно сэкономить память.

KudryashovAlexander replied:
Спасибо, буду иметь в виду

Helvetii951 commented on FakeNFT/Scenes /Cart/SucessPayView/SuccessPayViewController.swift:
```diff
@@ -0,0 +1,122 @@
+        ])
+    }
+    
+    @objc
+    private func backCataloge() {
+        guard let window = UIApplication.shared.windows.first else { fatalError("Invalid Configuration") }
+        let tabBar = TabBarViewController(viewModel: TabBarViewModel())
+        window.rootViewController = tabBar
+    }
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Обычно в таких случая пишут в логи. Локальные или удаленные.

Плюс очень рекомендую следить за правильностью именования свойтв и функций. Тут, например, опечатка.

KudryashovAlexander replied:
Хорошо, спасибо.
fatalError уберу

Helvetii951 commented on FakeNFT/Scenes /Cart/SucessPayView/SuccessPayViewController.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+    
+    private func updateContentSize() {
+        if traitCollection.verticalSizeClass == .compact {
+            topImageViewConstraint?.constant = 20
+            contenSize = CGSize(width: view.frame.width,
+                                height: 20 + successImageView.frame.height + 20 +
+                                        successLabel.frame.height + 20 + backCatalogeButton.frame.height + 16)
```
`20`, `16`

Желательно не применять [магические числа](https://bureau.ru/soviet/20230406/). Такие числовые переменные лучше выностить в свойства или структуры с понятными названиями.

KudryashovAlexander replied:
Понимаю, что так красивее. Постараюсь учесть в своем эпике, если успею.

Helvetii951 commented on FakeNFT/Scenes /Cart/CartWebView/CartWebViewController.swift:
```diff
@@ -0,0 +1,141 @@
+        return view
+    }()
+    
+    init(viewModel: CartWebViewModel) {
+        self.viewModel = viewModel
+        super.init(nibName: nil, bundle: nil)
+    }
+    
```
Рекомендую всегда внедрять зависимости по протоколу, а не конкретной реализацией. Так код будет гибче, легче будет тестировать.

[Хорошая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/)

KudryashovAlexander replied:
Вот тут у меня есть вопрос.
Пробовал закрыть все протоколами, но столкнулся со сложностью.
Так как в своем эпике использую наблюдаемые свойства `CartObservable`, я не знаю как их засунуть в протокол.  Xcode ругается что свойство, объявленное внутри протокола, не может иметь оболочку. Есть решение на [stackoverflow](https://stackoverflow.com/questions/57652925/propertywrappers-and-protocol-declaration), но код разрастается из-за добавления новых строк. Я использовал наблюдаемые свойства из-за меньшего размера по сравнение с замыканиями. При добавлении протоколов это преимущество теряется.
Если есть какое-то другое красивое решение, буду благодарен.

Helvetii951 replied:
> Вот тут у меня есть вопрос. Пробовал закрыть все протоколами, но столкнулся со сложностью. Так как в своем эпике использую наблюдаемые свойства `CartObservable`, я не знаю как их засунуть в протокол. Xcode ругается что свойство, объявленное внутри протокола, не может иметь оболочку. Есть решение на [stackoverflow](https://stackoverflow.com/questions/57652925/propertywrappers-and-protocol-declaration), но код разрастается из-за добавления новых строк. Я использовал наблюдаемые свойства из-за меньшего размера по сравнение с замыканиями. При добавлении протоколов это преимущество теряется. Если есть какое-то другое красивое решение, буду благодарен.

Да, обертки нельзя использовать в протоколах. Я бы использовал Combine. Но он выходит за пределы курса.
Такая реализация тоже пойдет. 

Рекомендую уже сейчас начинать знакомиться с реактивщиной. Концепция не очень простая, но супер сильная. Эти знания скоро будут необходимы в связке SwiftUI + Combine. Плюс такие знания станут преимуществом при трудоустройстве.

KudryashovAlexander replied:
Понял, спасибо за развернутый ответ.

Helvetii951 commented on FakeNFT/Generated/Strings+Generated.swift:
```diff
@@ -11,6 +11,10 @@ import Foundation
 // swiftlint:disable nesting type_body_length type_name vertical_whitespace_opening_braces
 public enum L {
   public enum Cart {
+    /// Вернуться в каталог
+    public static let backCataloge = L.tr("Localizable", "cart.backCataloge", fallback: "Вернуться в каталог")
+    /// Отмена
+    public static let cancel = L.tr("Localizable", "cart.cancel", fallback: "Отмена")
     /// Корзина пуста
```
Повторюсь, все сгенеренные файлы лучше выносить в .gitignore. Они не нужны в репозитории.

KudryashovAlexander replied:
Ссори, забыл, поправлю

KudryashovAlexander committed:
Поправлено после ревью:
1) доработано отображение progressHud. До этого крутился бесконечно после успешной оплаты и перехода в таббар
2) Добавлены константы на successView и WebView
